### PR TITLE
ci: Include what you use

### PR DESCRIPTION
**Pull request by andiwand**
In this PR I set up a basic workflow to run [IWYU](https://github.com/include-what-you-use/include-what-you-use) and a filtering mechanism for its output.

My hope is that it could lead to accelerated build times and cleaner include flow.

### CODE REVIEW

File: .github/workflows/iwyu.yml [hunk 1]
```diff
@@ -0,0 +1,70 @@
+name: IWYU
+
+on:
+  push: # TODO remove
```
Comment (andiwand) [inline .github/workflows/iwyu.yml:4]: TODO this should go

File: .github/workflows/analysis.yml [hunk 2]
```diff
@@ -1,16 +1,6 @@
...
-  push:
-    branches:
-      - main
-  pull_request:
-    branches:
-      - main
-      - 'release/**'
-      - 'develop/**'
-    paths-ignore:
-      - "docs/**"
```
Comment (andiwand) [inline .github/workflows/analysis.yml:13]: revert

File: .github/workflows/builds.yml [hunk 3]
```diff
@@ -1,14 +1,6 @@
...
-on:
-  push:
-  pull_request:
-    branches:
-      - main
-      - 'release/**'
-      - 'develop/**'
-    paths-ignore:
-      - "docs/**"
+on:
```
Comment (andiwand) [inline .github/workflows/builds.yml:3]: revert

File: .github/workflows/checks.yml [hunk 4]
```diff
@@ -1,12 +1,6 @@
 name: Checks
 
 on:
-  push:
-  pull_request:
-    branches:
-      - main
-      - 'release/**'
-      - 'develop/**'
```
Comment (andiwand) [inline .github/workflows/checks.yml:9]: revert

File: iwyu/filter.yaml [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  - "namespace Eigen {"
+  # ignore boost
+  - "^(- )?#include <boost/"
+  # ignore HEP
+  - "^#include <CLHEP/"
+  - "^#include <TGeo"
+  - "^(- )?#include \"TGeo"
+  - "^(- )?#include \"TView"
+  - "^(- )?#include [<\"]G4"
+  # include touch ipp
```
Comment (andiwand) [inline iwyu/filter.yaml:17]:
```suggestion
  # don't touch ipp
```

File: .github/workflows/iwyu.yml [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        run: apt-get update && apt-get install -y git-lfs llvm-dev libclang-dev clang
+
+      - name: Get IWYU source
+        uses: actions/checkout@v3
+        with:
+          repository: include-what-you-use/include-what-you-use
+          ref: clang_14
+          path: iwyu
+
+      - name: Build IWYU
```
Comment (paulgessinger) [inline .github/workflows/iwyu.yml:27]: Would it be worth adding a docker image for this? How long does the compilation take? I suppose on schedule it doesn't really matter.

File: iwyu/filter.yaml
Comment (paulgessinger) [inline iwyu/filter.yaml:1]: Same with this one

File: .github/workflows/iwyu.yml (hunk 6 above)
Comment (andiwand) [reply .github/workflows/iwyu.yml:27]: yeah I think for now it is fine. the build is actually quite fast

File: iwyu/filter.py
[the remark this answers is not in the document]
Comment (andiwand) [reply iwyu/filter.py:1]: will do

**Review by paulgessinger — APPROVED** [review]
Let's go then!

**Comment by andiwand** [comment]
@paulgessinger sorry had to commit one more time after kicking the ci

File: CI/iwyu/filter.yaml [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  - "^#include <new>"
+  - "^#include <bits/"
+  - "^#include <ext/"
+  # ignore Eigen
+  - "^(- )?#include [<\"](src|Eigen)/"
+  - "namespace Eigen {"
+  # ignore boost
+  - "^(- )?#include <boost/"
+  # don remove ipp
+  - "^- #include [<\"].*\\.ipp": ""
```
Comment (andiwand) [inline CI/iwyu/filter.yaml:12]:
```suggestion
  - "^- #include [<\"].*\\.ipp"
```

**Review by paulgessinger — APPROVED** [review]

**Pull request merged**